x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java

```diff
@@ -177,13 +186,68 @@ public void getEngine(String resourceName, ActionListener<Engine> listener) {
         }));
     }
 
+    private static String getEngineAliasName(Engine engine) {
+        return ENGINE_ALIAS_PREFIX + engine.name();
+    }
+
     /**
      * Creates or updates the {@link Engine} in the underlying index.
      *
      * @param engine The engine object.
      * @param listener The action listener to invoke on response/failure.
      */
     public void putEngine(Engine engine, ActionListener<IndexResponse> listener) {
+        validateIndices(engine);
+
+        createOrUpdateAlias(engine, listener.delegateFailure((l, aliasResponse) -> { updateEngine(engine, l); }));
+
+    }
+
+    private void createOrUpdateAlias(Engine engine, ActionListener<AcknowledgedResponse> listener) {
+
+        final Metadata metadata = clusterService.state().metadata();
+        final String engineAliasName = getEngineAliasName(engine);
+
+        IndicesAliasesRequestBuilder requestBuilder = null;
+        if (metadata.hasAlias(engineAliasName)) {
+            Set<String> currentAliases = metadata.aliasedIndices(engineAliasName)
+                .stream()
+                .map(index -> index.getName())
+                .collect(Collectors.toSet());
+            Set<String> targetAliases = Set.of(engine.indices());
+
+            if (currentAliases.equals(targetAliases)) {
```

Comment:
We came up with a subtle edge case when developing this in the Rails app: 

- User creates an Engine with `index1` and `index2`. 
- User deletes `index2`. It is automatically removed from the alias. 
- User re-creates `index2`. The alias is not automatically updated. 
- User adds `index2` to the engine. Since `index2` was already associated with the engine no action is taken and the index is not added back to the alias. 

The workaround here I think is to always add all the indices to the alias on every PUT request. It's a cheap and atomic operation.